packages/portfolio-contract/tools/plan-solve.ts

```diff
@@ -388,72 +406,145 @@ export const solveRebalance = async (
-      .map(f => [f.edge.id, { ...f, srcChain: chainOf(f.edge.src) }]),
-  );
-  const prioritized = [] as AnnotatedFlow[];
+  const scheduled = new Set<FlowIndex>();
+  const order: StepOrder = [];
 
   // Maintain last chosen originating chain to group sequential operations.
   let lastChain: string | undefined;
 
-  while (pendingFlows.size) {
-    // Find flows that can be executed based on current supplies.
-    const candidates = [...pendingFlows.values()].filter(
-      f => (supplies.get(f.edge.src) || 0) >= f.flow,
-    );
+  while (scheduled.size < flows.length) {
+    /** Flows that can execute with current available supply */
+    const candidates: { flowIdx: FlowIndex; flow: SolvedEdgeFlow }[] = [];
+    for (let i = 0; i < flows.length; i += 1) {
+      if (scheduled.has(i)) continue;
+      const flow = flows[i];
+      const srcSupply = available.get(flow.edge.src) || 0;
+      if (srcSupply >= flow.flow) candidates.push({ flowIdx: i, flow });
+    }
 
+    // Fail upon encountering deadlock.
     if (!candidates.length) {
-      // Deadlock detected: cannot schedule remaining flows.
-      // This indicates a solver bug or rounding error that produced an infeasible flow.
-      const diagnostics = [...pendingFlows.values()].map(f => {
-        const srcSupply = supplies.get(f.edge.src) || 0;
-        const shortage = f.flow - srcSupply;
-        return `${f.edge.id}: ${f.edge.src}(${srcSupply}) -> ${f.edge.dest} needs ${f.flow} (short ${shortage})`;
+      const remaining = partialMap(flows, (f, idx) => {
+        if (scheduled.has(idx)) return;
+        const { id, src, dest } = f.edge;
+        const srcSupply = available.get(src);
+        return `${idx}: ${id} ${src}(${srcSupply}) -> ${dest} needs ${f.flow}`;
       });
-      failUnsolvable(
-        X`Scheduling deadlock: no flows can be executed. Remaining flows:\n${diagnostics.join('\n')}`,
-      );
+      throw Fail`Scheduling deadlock: no flows can be executed. Remaining:\n${remaining.join('\n')}`;
     }
-    // Prefer continuing with lastChain if possible.
+
+    // Prefer continuing with lastChain if possible
     const fromSameChain = lastChain
-      ? candidates.filter(c => c.srcChain === lastChain)
+      ? candidates.filter(c => chainOf(c.flow.edge.src) === lastChain)
       : undefined;
     const chosenGroup = fromSameChain?.length ? fromSameChain : candidates;
 
-    // Pick deterministic smallest edge id within chosen group.
-    chosenGroup.sort((a, b) => naturalCompare(a.edge.id, b.edge.id));
-    const chosen = chosenGroup[0];
-    prioritized.push(chosen);
-    replaceOrInit(supplies, chosen.edge.src, (old = 0) => old - chosen.flow);
-    replaceOrInit(supplies, chosen.edge.dest, (old = 0) => old + chosen.flow);
-    pendingFlows.delete(chosen.edge.id);
-    lastChain = chosen.srcChain;
+    // Pick deterministic smallest edge id within chosen group
+    chosenGroup.sort((a, b) => naturalCompare(a.flow.edge.id, b.flow.edge.id));
```

Comment:
This uses a different sort order than the outer loop. If we keep using the flowID sort order, then the flows order can describe the preference order.  Not required to change now.